### PR TITLE
Publish refactor

### PR DESCRIPTION
**Pull request by Xarthisius**
Counterpart PR to https://github.com/whole-tale/gwvolman/pull/93. This PR:

* Adds `PUT /tale/:id/publish` as a generic publishing endpoint, that only takes a `repository` as an argument
* Allows to configure repositories where Tales can be deposited.
* Provides `GET /repository` per #368 

How to test?
1. Deploy along with https://github.com/whole-tale/gwvolman/pull/93.
1. Create a Tale
1. Using swagger hit `GET /repository` and confirm it returns an empty array
1. Add your `sandbox.zenodo.org` api key via https://girder.local.wholetale.org/#ext_keys
1. Using swagger hit `GET /repository` and confirm it returns `["sandbox.zenodo.org"]`
1. Using swagger publish your Tale with `repository` == "sandbox.zenodo.org"
1. Authenticate with DataONE stage-2 via https://girder.local.wholetale.org/#ext_keys
1. Using swagger hit `GET /repository` and confirm it returns `["https://.../d1/mn", "sandbox.zenodo.org"]`
1. Using swagger publish your Tale with `repository` == "https://dev.nceas.ucsb.edu/knb/d1/mn"

### CODE REVIEW

File: server/rest/tale.py [hunk 1]
```diff
@@ -634,3 +636,56 @@ def _extractZipPayload():
                 # unzipping an untrusted content. What could possibly go wrong...?
                 z.extractall(path=temp_dir)
         return temp_dir, manifest_file, manifest, environment
+
+    @access.user(scope=TokenScope.DATA_WRITE)
+    @filtermodel(model=Job)
+    @autoDescribeRoute(
+        Description("Publish a tale to a repository running Metacat")
```
Comment (craig-willis) [inline server/rest/tale.py]: Metacat!?

Comment (Xarthisius) [reply server/rest/tale.py]: Copy and paste :blush: Fixed in 9a1f3f0

**Comment by bodom0015** [comment]
> Noting that DataONE account config doesn't yet work via Dashboard (whole-tale/dashboard#563) so I had to use the ext_keys.

You should now be able to connect a DataONE account from the dashboard, as that PR you linked had been merged already. Is this not functioning correctly for you?

**Comment by craig-willis** [comment]
No, it didn't work for me.  It looked fine in the UI (state was connected, etc), but on the backend it looked like:

```
 {
  'access_token': '',
  'provider': 'dataonestage2',
  'resource_server': 'cn-stage-2.test.dataone.org',
  'token_type': 'dataone-pre'
]
```

I assumed from the comment in your Dashboard PR that this was a known issue.  It worked via Kacper's ext_keys Girder UI component.

**Comment by Xarthisius** [comment]
@bodom0015 DataONE needs a bit of special handling. I tried to describe the details in https://github.com/whole-tale/girder_wholetale/pull/364 Please let me know if I did a poor job and if you have any questions.

**Comment by bodom0015** [comment]
@craig-willis I'm not sure to which PR you are referring, but last I knew this was working.

From https://github.com/whole-tale/dashboard/pull/563#issuecomment-553097671:
> Finally, I've added support for provider.type=='bearer' and provider.type=='dataone' by redirecting to provider.url on Connect. To Disconnect, we call the /revoke endpoint just as before. Note that the Connect button must be disabled when provider.state=='authorized' or provider.state=='preauthorized', since this changes the provider.url to a different (invalid) value.

I don't recall this coming up during the UI review... please file a bug if you haven't already.

**Comment by craig-willis** [comment]
OK, I was referring to the note in the issue description, not that comment. Apparently I didn't test this -- or didn't confirm the token was being stored in Girder.

I filed https://github.com/whole-tale/dashboard/issues/573.

**Comment by bodom0015** [comment]
Ah, I guess I forgot to remove that note from the test case ¯\_(ツ)_/¯

... or even go back and update it at all, apparently

**Comment by bodom0015** [comment]
Without any test expectations I am not sure what it means to work or not to work. I get a job back, but the Tale's `publishInfo` remains empty.

I also get a nasty stack trace back in the `gwvolman` logs:
```
User information: uid=0 euid=0 gid=0 egid=0

  uid=uid, euid=euid, gid=gid, egid=egid,
[2019-11-22 18:21:38,440: ERROR/ForkPoolWorker-3] Task gwvolman.tasks.publish[b1e66c5c-9c27-4b2b-bb65-85744bcaaa12] raised unexpected: HTTPError('HTTP error 500: GET http://girder:8080/api/v1//tale/5dd8236f63d3c1f398dfa073/manifest',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/celery/app/trace.py", line 382, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/girder_worker/task.py", line 148, in __call__
    results = super(Task, self).__call__(*_t_args, **_t_kwargs)
  File "/usr/local/lib/python3.5/dist-packages/celery/app/trace.py", line 641, in __protected_call__
    return self.run(*args, **kwargs)
  File "/gwvolman/gwvolman/tasks.py", line 505, in publish
    dataone_node=repository,
  File "/gwvolman/gwvolman/lib/dataone/publish.py", line 41, in __init__
    super().__init__(gc, tale_id, token, draft=draft, job_manager=job_manager)
  File "/gwvolman/gwvolman/lib/publish_provider.py", line 30, in __init__
    self.manifest = self.gc.get("/tale/{}/manifest".format(tale_id))
  File "/usr/local/lib/python3.5/dist-packages/girder_client/__init__.py", line 503, in get
    return self.sendRestRequest('GET', path, parameters, jsonResp=jsonResp)
  File "/usr/local/lib/python3.5/dist-packages/girder_client/__init__.py", line 497, in sendRestRequest
    response=result)
requests.exceptions.HTTPError: HTTP error 500: GET http://girder:8080/api/v1//tale/5dd8236f63d3c1f398dfa073/manifest

```

**Comment by Xarthisius** [comment]
@bodom0015 could you call `GET /tale/5dd8236f63d3c1f398dfa073/manifest`
through swagger and see why it fails.

Test expectation is a published Tale. Login to sandbox.zenodo.org, click on uploads and see if it's there.

**Comment by bodom0015** [comment]
@Xarthisius this is probably because I created the Tale through Swagger UI and we don't have enough server-side validation there:
```json
{
  "message": "KeyError: KeyError('_modelType',)",
  "trace": [
    "<FrameSummary file /girder/girder/api/rest.py, line 630 in endpointDecorator>",
    "<FrameSummary file /girder/girder/api/rest.py, line 1210 in GET>",
    "<FrameSummary file /girder/girder/api/rest.py, line 967 in handleRoute>",
    "<FrameSummary file /girder/girder/api/describe.py, line 709 in wrapped>",
    "<FrameSummary file /girder/plugins/wholetale/server/rest/tale.py, line 467 in generateManifest>",
    "<FrameSummary file /girder/plugins/wholetale/server/lib/manifest.py, line 47 in __init__>",
    "<FrameSummary file /girder/plugins/wholetale/server/lib/manifest.py, line 222 in add_tale_records>",
    "<FrameSummary file /girder/plugins/wholetale/server/lib/manifest.py, line 287 in _parse_dataSet>"
  ],
  "type": "internal"
}
```

This is likely unrelated to the changes in this PR, but I can re-test to confirm

**Comment by bodom0015** [comment]
@Xarthisius I went ahead and created a Tale with the current (EmberJS) Dashboard instead and publishing to DataONE produced a new entry in `publishInfo` as expected. When attempting to publish to Zenodo, I get a different error:
```
[2019-11-22 19:48:35,869: INFO/MainProcess] Received task: gwvolman.tasks.publish[ab9eb419-4436-4852-8d2d-39c7f97ad0b7]  
[2019-11-22 19:48:35,915: ERROR/ForkPoolWorker-2] Task gwvolman.tasks.publish[ab9eb419-4436-4852-8d2d-39c7f97ad0b7] raised unexpected: AttributeError("'NoneType' object has no attribute 'strip'",)
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/celery/app/trace.py", line 382, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/girder_worker/task.py", line 148, in __call__
    results = super(Task, self).__call__(*_t_args, **_t_kwargs)
  File "/usr/local/lib/python3.5/dist-packages/celery/app/trace.py", line 641, in __protected_call__
    return self.run(*args, **kwargs)
  File "/gwvolman/gwvolman/tasks.py", line 518, in publish
    provider.publish()
  File "/gwvolman/gwvolman/lib/zenodo.py", line 199, in publish
    deposition = self.create_deposition()
  File "/gwvolman/gwvolman/lib/zenodo.py", line 141, in create_deposition
    zenodo_metadata = self.get_tale_metadata()
  File "/gwvolman/gwvolman/lib/zenodo.py", line 123, in get_tale_metadata
    self.manifest["schema:description"]
  File "/gwvolman/gwvolman/lib/zenodo.py", line 53, in _render_description
    html = parser.convert(description)
  File "/usr/local/lib/python3.5/dist-packages/markdown/core.py", line 252, in convert
    if not source.strip():
AttributeError: 'NoneType' object has no attribute 'strip'
```

And here is the offending Tale:
```json
  {
    "_accessLevel": 2,
    "_id": "5dd83a2f0433fad654f3fe1c",
    "_modelType": "tale",
    "authors": [],
    "category": "science",
    "config": {},
    "copyOfTale": null,
    "created": "2019-11-22T19:42:39.116000+00:00",
    "creatorId": "5dc1f1088573ab02472175d9",
    "dataSet": [],
    "dataSetCitation": [],
    "description": null,
    "folderId": "5dd83a2f0433fad654f3fe1e",
    "format": 7,
    "icon": "https://raw.githubusercontent.com/whole-tale/jupyter-base/master/squarelogo-greytext-orangebody-greymoons.png",
    "iframe": true,
    "illustration": "https://raw.githubusercontent.com/whole-tale/dashboard/master/public/images/demo-graph2.jpg",
    "imageId": "5dc1f0ea8573ab02472175d4",
    "imageInfo": {
      "jobId": "5dd83a2f0433fad654f3fe21",
      "status": 2
    },
    "licenseSPDX": "CC-BY-4.0",
    "narrative": [],
    "narrativeId": "5dc5d8048573ab02472175e6",
    "public": false,
    "publishInfo": [],
    "status": 1,
    "title": "This is a Brand New Tale that was created with the EmberJS WT Dashboard",
    "updated": "2019-11-22T19:42:39.347000+00:00",
    "workspaceId": "5dd83a2f0433fad654f3fe1d"
  }
```

**Comment by Xarthisius** [comment]
The latter is failing because description is `None`. I can only fix that partially by providing better error message, but for Zenodo description is a required field AFAIR, so it's gonna fail either way.

**Comment by Xarthisius** [comment]
The former is caused by some quirk in the `dataSet`. Does item `5c4887389759c200017b230e` actually exist?

**Comment by bodom0015** [comment]
Yep, you were correct that it was mad because the `description` was null. Ideally, this error message is more meaningful. We could probably just default `description` to an empty string (or a placeholder value of "Please enter a description", etc) instead of null, and then it might not fail by default.

After fixing that, I got a few more errors from the actual Zenodo API. One of these errors being that they actually _do_ validate the author ORCIDs, so we should eventually do this as well.

Now it has plopped out this, but the DOI (https://doi.org/10.5072/zenodo.421415) does not seem to lead anywhere:
```
[2019-11-22 19:58:39,141: WARNING/ForkPoolWorker-2] {'doi': '10.5072/zenodo.421415', 'links': {'doi': 'https://doi.org/10.5072/zenodo.421415', 'record_html': 'https://sandbox.zenodo.org/record/421415', 'record': 'https://sandbox.zenodo.org/api/records/421415', 'bucket': 'https://sandbox.zenodo.org/api/files/59112b85-c5b9-4e96-a27f-13c9de664010', 'conceptbadge': 'https://sandbox.zenodo.org/badge/doi/10.5072/zenodo.421414.svg', 'latest': 'https://sandbox.zenodo.org/api/records/421415', 'badge': 'https://sandbox.zenodo.org/badge/doi/10.5072/zenodo.421415.svg', 'latest_html': 'https://sandbox.zenodo.org/record/421415', 'conceptdoi': 'https://doi.org/10.5072/zenodo.421414'}, 'title': 'This is a Brand New Tale', 'submitted': True, 'modified': '2019-11-22T19:58:38.711512+00:00', 'state': 'done', 'conceptdoi': '10.5072/zenodo.421414', 'metadata': {'doi': '10.5072/zenodo.421415', 'keywords': ['Science', 'Tale'], 'references': [], 'description': '<div>Markdown Editor</div>', 'creators': [{'orcid': '0000-0003-4889-5010', 'name': 'Lambert, Mike'}], 'license': 'CC-BY-4.0', 'title': 'This is a Brand New Tale', 'publication_type': 'other', 'upload_type': 'publication', 'notes': 'Run this Tale on Whole Tale by clicking <a href="https://data.wholetale.org/api/v1/integration/zenodo?doi=10.5072%2Fzenodo.421415">here</a>.', 'publication_date': '2019-11-22', 'access_right': 'open', 'prereserve_doi': {'doi': '10.5072/zenodo.421415', 'recid': 421415}, 'communities': [{'identifier': 'zenodo'}]}, 'created': '2019-11-22T19:58:35.657637+00:00', 'conceptrecid': '421414', 'doi_url': 'https://doi.org/10.5072/zenodo.421415', 'id': 421415, 'owner': 30967, 'files': [{'filesize': 7762, 'links': {'self': 'https://sandbox.zenodo.org/api/deposit/depositions/421415/files/067fa5be-5991-4a40-89af-7f1ac85ae33b', 'download': 'https://sandbox.zenodo.org/api/files/59112b85-c5b9-4e96-a27f-13c9de664010/5dd83a2f0433fad654f3fe1c.zip'}, 'id': '067fa5be-5991-4a40-89af-7f1ac85ae33b', 'filename': '5dd83a2f0433fad654f3fe1c.zip', 'checksum': '57697f52b2c044b108936030f12533cf'}], 'record_id': 421415}
[2019-11-22 19:58:39,197: INFO/ForkPoolWorker-2] Task gwvolman.tasks.publish[835b4c2b-555a-4169-a461-05f3ed619e6c] succeeded in 4.142199752852321s: None
```

I am logged in on sandbox.zenodo.org, but I don't see a place for me to check my uploads or anything.... so.... I guess I'm still not sure if it worked or not

**Comment by Xarthisius** [comment]
Yeah, DOIs for sandbox don't resolve. After logging in to sandbox go here: https://sandbox.zenodo.org/deposit

**Comment by bodom0015** [comment]
Ah... there they are. LGTM

Sorry, but I have not used many of these external systems before

**Pull request merged**